File: 15_orderbook/main.cpp
#include <iostream>
#include <vector>

using namespace std;

#define NUM_ENTRIES 1000000
#define MAX_DEPTH 1 << 24

#if __linux__
#include <linux/version.h>
#include <sys/mman.h>
#include <unistd.h>
#endif

#ifdef DEBUG
#define DEBUG_PRINT(...) fprintf(stderr, __VA_ARGS__)
#else
#define DEBUG_PRINT(...) ;
#endif

struct Order {
  int price, size;
};

class OrderBook {
  std::vector<Order> orders;

 public:
  OrderBook() {
    orders.reserve(MAX_DEPTH);
    DEBUG_PRINT("reserved orders with %d\n", MAX_DEPTH);
  }

  void add(int price, int size) {
    Order order = Order{.price = price, .size = size};

    // if first order
    if (orders.size() == 0) {
      orders.push_back(order);
      return;
    }

    // check with the last order
    if (orders.size() > 1) {
      DEBUG_PRINT("checking worst ask\n");
      auto last = *(orders.end() - 1);
      // if bigger than the last we can just insert at the back
      if (price > last.price || (price == last.price && size >= last.size)) {
        DEBUG_PRINT("inserting at back\n");
        orders.push_back(order);
        return;
      }
      // if same price as last but smaller size, we find the pos from back
      else if (price == last.price) {
        DEBUG_PRINT("same price as last, reversing to find pos\n");
        for (int i = orders.size() - 2; i >= 0; i--)
          if (price == orders[i].price && size < orders[i].size)
            continue;
          else {
            DEBUG_PRINT("inserting at: %d\n", i + 1);
            orders.insert(orders.begin() + i + 1, order);
            return;
          }
      }
    }

    DEBUG_PRINT("checking best_ask\n");
    auto best_ask = orders[0];

    // smaller price, we insert in front
    if (price < best_ask.price ||
        // if equal price and size is correct, we also insert front
        (price == best_ask.price && size < best_ask.size)) {
      orders.insert(orders.begin(), order);
      return;
    }
    // if same price but larger size then we find the next position for the size
    else if (price == best_ask.price) {
      for (int i = 1; i < orders.size(); i++) {
        if (price == orders[i].price && size > orders[i].size)
          continue;
        else {
          orders.insert(orders.begin() + i, order);
          return;
        }
      }
    }

    // bigger price, so we advance and find the position
    DEBUG_PRINT("bigger price\n");
    for (int i = 1; i < orders.size(); i++)
      if (price > orders[i].price)
        continue;
      else if (price == orders[i].price && size > orders[i].size)
        continue;
      else {
        DEBUG_PRINT("inserting at %d\n", i);
        orders.insert(orders.begin() + i, order);
        return;
      }

    // if not, worst case we add at the back
    orders.push_back(order);
  }

  void remove(int pos) {
    orders.erase(orders.begin() + min(orders.size() - 1, (size_t)pos));
  }

  int buy(int shares) {
    int cost = 0;
    while (shares > 0 && orders.size() > 0) {
      auto it = orders.begin();
      int size = min(shares, it->size);
      shares -= size;
      cost += size * it->price;
      DEBUG_PRINT("it: (%d, %d), size: %d, cost: %d\n", it->price, it->size,
                  size, cost);
      it->size -= size;
      if (it->size == 0)
        orders.erase(it);
    }
    return cost;
  }

  void print() {
    for (auto it = orders.begin(); it != orders.end(); it++) {
      DEBUG_PRINT("(%d, %d)", it->price, it->size);
      if (it != orders.end() - 1)
        DEBUG_PRINT(", ");
    }
    if (orders.size() > 0)
      DEBUG_PRINT("\n");
  }
};

int i = 0;
char c;
OrderBook orderbook;

#if __linux__
#define OP_ADD_GET_PRICE 0
#define OP_ADD_GET_SIZE 1
#define OP_REMOVE_POS 2
#define OP_BUY_SHARES 3

int main() {
  off_t fsize = lseek(0, 0, SEEK_END);
  char* buffer =
      (char*)mmap(0, fsize, PROT_READ, MAP_PRIVATE | MAP_POPULATE, 0, 0);

  uint8_t state = OP_ADD_GET_PRICE;
  std::string token;
  int price, size, pos, shares;

  for (int i = 0; i < fsize; i++) {
    switch (buffer[i]) {
      case '+':
        state = OP_ADD_GET_PRICE;
        i++;
        continue;
      case '-':
        state = OP_REMOVE_POS;
        i++;
        continue;
      case '=':
        state = OP_BUY_SHARES;
        i++;
        continue;
      default:
        break;
    }

    switch (state) {
      case OP_ADD_GET_PRICE:
        if (buffer[i] == ' ') {
          price = stoi(token);
          token = "";
          state = OP_ADD_GET_SIZE;
          continue;
        }
        token += buffer[i];
        break;
      case OP_ADD_GET_SIZE:
        if (buffer[i] == '\n') {
          size = stoi(token);
          token = "";
          orderbook.add(price, size);
          continue;
        }
        token += buffer[i];
        break;
      case OP_REMOVE_POS:
        if (buffer[i] == '\n') {
          pos = stoi(token);
          token = "";
          orderbook.remove(pos);
          continue;
        }
        token += buffer[i];
        break;
      case OP_BUY_SHARES:
        if (buffer[i] == '\n') {
          shares = stoi(token);
          token = "";
          orderbook.buy(shares);
          continue;
        }
        token += buffer[i];
        break;
    }
  }

  // buy 1000 shares and report the total cost
  cout << orderbook.buy(1000) << endl;

  return 0;
}
#endif

#if __APPLE__
int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  for (; i < NUM_ENTRIES; i++) {
    cin >> c;
    // DEBUG_PRINT("i: %d, got: %c", i, c);
    switch (c) {
      case '+':
        int size, price;
        cin >> price >> size;
        // DEBUG_PRINT(" %d %d\n", price, size);
        orderbook.add(price, size);
        break;

      case '-':
        int pos;
        cin >> pos;
        // DEBUG_PRINT(" %d\n", pos);
        orderbook.remove(pos);
        break;

      case '=':
        cin >> size;
        // DEBUG_PRINT(" %d\n", size);
        orderbook.buy(size);
        break;
    }
    // orderbook.print();
  }

  // buy 1000 shares and report the total cost
  cout << orderbook.buy(1000) << endl;
  return 0;
}
#endif

File: 15_orderbook/main.rs
#[link(name = "c")]
extern {
    fn mmap(addr: *mut u8, len: usize, prot: i32, flags: i32, fd: i32, offset: i64) -> *mut u8;
    // fn madvise(addr: *mut u8, len: usize, advice: i32) -> i32;
    fn __errno_location() -> *const i32;
    fn lseek(fd: i32, offset: i64, whence: i32) -> i64;
    fn open(path: *const u8, oflag: i32) -> i32;
}

#[allow(dead_code)]
unsafe fn mmap_stdin<'a>() -> &'a [u8] {
    mmap_fd(0)
}

#[allow(dead_code)]
unsafe fn mmap_path<'a>(path: &str) -> &'a [u8] {
    let mut path2 = vec![];
    path2.extend_from_slice(path.as_bytes());
    path2.push(0);
    let fd = open(path2.as_ptr(), 0);
    if fd == -1 {
        panic!("open failed, errno {}", *__errno_location());
    }
    mmap_fd(fd)
}

unsafe fn mmap_fd<'a>(fd: i32) -> &'a [u8] {
    let seek_end = 2;
    let size = lseek(fd, 0, seek_end);
    if size == -1 {
        panic!("lseek failed, errno {}", *__errno_location());
    }
    let prot_read = 0x01;
    let map_private = 0x02;
    let map_populate = 0x08000;
    // let madv_hugepage = 0x0E; 
    // let ptr = madvise(0 as _, size as usize, madv_hugepage);
    // if ptr as isize == -1 {
    //     panic!("madvise failed, errno {}", *__errno_location());
    // }
    let ptr = mmap(0 as _, size as usize, prot_read, map_private | map_populate, fd, 0);
    if ptr as isize == -1 {
        panic!("mmap failed, errno {}", *__errno_location());
    }
    std::slice::from_raw_parts(ptr, size as usize)
}

#[derive(Debug, PartialOrd, Ord, PartialEq, Eq)]
struct Order {
    price: u32,
    size: u32,
}

struct Orderbook {
    // orders: Vec<Order>
    orders: BTreeSet<Order>
}

impl Orderbook {
    fn new() -> Self {
        Self { orders: Vec::with_capacity(u32::MAX as usize) }
    }

    fn add(&mut self, price: u32, size: u32) {
        let order = Order { price, size };
        // println!("add: {} {}", price, size);

        // if first order
        if self.orders.len() == 0 {
            self.orders.push(order);
            return;
        }

        // check with the last order
        if self.orders.len() > 1 {
            let last = &self.orders[self.orders.len() - 1];

            // if bigger than the last we can just insert at the back
            if price > last.price || (price == last.price && size >= last.size) {
                self.orders.push(order);
                return;
            }

            // if same price as last but smaller size, we find the pos from back
            if price == last.price {
                for (i, o) in self.orders.iter().rev().enumerate() {
                    if price == o.price && size < o.size {
                        continue 
                    } else {
                        self.orders.insert(i + 1, order);
                        return;
                    }
                }
            }
        }

        let best_ask = &self.orders[0];

        // smaller price, we insert in front
        if price < best_ask.price || (price == best_ask.price && size < best_ask.size) {
            self.orders.insert(0, order);
            return;
        }

        // if same price but larger size then we find the next position for the size
        if price == best_ask.price {
            for i in 1..self.orders.len() {
                if price == self.orders[i].price && size > self.orders[i].size {
                    continue;
                } else {
                    self.orders.insert(i, order);
                    return;
                }
            }
        }

        // bigger price, so we advance and find the position
        for i in 1..self.orders.len() {
            if price > self.orders[i].price {
                continue;
            } else if price == self.orders[i].price && size > self.orders[i].size {
                continue;
            } else {
                self.orders.insert(i, order);
                return;
            }
        }

        // if not, worst case we add at the back
        self.orders.push(order);
    }

    fn remove(&mut self, pos: usize) {
        // println!("remove: {}", pos);
        self.orders.remove(std::cmp::min(pos, self.orders.len() - 1));
    }

    fn buy(&mut self, mut shares: u32) -> u32 {
        // println!("buy: {}", shares);
        let mut cost = 0;
        while shares > 0 && self.orders.len() > 0 {
            let order = &mut self.orders[0];
            let size = std::cmp::min(shares, order.size);
            shares -= size;
            cost += size * order.price;
            order.size -= size;
            if order.size == 0 {
                self.orders.remove(0);
            }
        }
        cost
    }

    #[allow(unused)]
    fn print(&self) {
        println!("{:?}", self.orders);
    }
}

enum Action {
    Add { price: u32, size: u32 },
    Remove { pos: usize },
    Buy { shares: u32 }
}

impl Action {
    fn from_str(string: &str) -> Self {
        let mut splits = string.trim().split(' ');
        match splits.next().unwrap() {
            "+" => {
                let price = splits.next().unwrap().parse().unwrap();
                let size = splits.next().unwrap().parse().unwrap();
                Self::Add { price, size }
            },
            "-" => {
                let pos = splits.next().unwrap().parse().unwrap();
                Self::Remove { pos }
            },
            "=" => {
                let shares = splits.next().unwrap().parse().unwrap();
                Self::Buy { shares }
            },
            _ => panic!("unsupported operation"),
        }
    }
}

fn main() {
    let mut orderbook = Orderbook::new();
    let buffer = std::str::from_utf8(unsafe { mmap_stdin() }).unwrap();
    let parts = buffer.split("\n");

    for part in parts {
        if part == "" {
            continue
        }
        match Action::from_str(&part) {
            Action::Add { price, size } => orderbook.add(price, size),
            Action::Remove { pos } => orderbook.remove(pos),
            Action::Buy { shares } => { orderbook.buy(shares); },
        };
        // orderbook.print();
    }

    println!("{}", orderbook.buy(1000));
}